srcts/src/components/darkModeSwitch.ts

```diff
@@ -0,0 +1,298 @@
+        transition-delay: 0s;
+      }
+    `,
+    css`
+      :host {
+        display: inline-block;
+
+        /* We control the stroke size manually here. We don't want it getting so
+        small its not visible but also not so big it looks cartoonish */
+        --stroke-w: clamp(1px, 0.1em, 6px);
+      }
+
+      button {
+        /* This is needed to let the svg use the em sizes */
+        font-size: inherit;
+
+        /* Make sure the button is fully centered */
+        display: grid;
+        place-content: center;
+
+        /* A little bit of padding to make it easier to press */
+        padding: var(--size-xxs);
+        background: none;
+        border: none;
+        aspect-ratio: 1;
+        border-radius: 50%;
+        cursor: pointer;
+        touch-action: manipulation;
+        -webkit-tap-highlight-color: transparent;
+        outline-offset: var(--size-xxs);
+
+        transform: translateY(var(--vertical-correction));
+        margin-block-end: var(--vertical-correction);
+      }
+
+      /*
+      button:is(:hover, :focus-visible) {
+        background: var(--surface-4);
+      }
+      */
+
+      button > svg {
+        height: var(--size);
+        width: var(--size);
+        stroke-linecap: round;
+        overflow: visible;
+      }
+
+      svg line,
+      svg circle {
+        vector-effect: non-scaling-stroke;
+      }
+    `,
+  ];
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function, @typescript-eslint/no-unused-vars
+  onChangeCallback: (x: boolean) => void = (x: boolean) => {};
+
+  // onValueChange = makeValueChangeEmitter(this, this.id);
+
+  connectedCallback(): void {
+    super.connectedCallback();
+
+    this.themeAttribute =
+      this.getAttribute("theme-attribute") || "data-shinytheme";
+
+    if (window.matchMedia("(prefers-color-scheme: dark)").matches) {
+      this.themeValue = "dark";
+    }
+
+    this.setPreference();
+
+    // Sync with system changes
+    window
+      .matchMedia("(prefers-color-scheme: dark)")
+      .addEventListener("change", ({ matches: isDark }) => {
+        this.themeValue = isDark ? "dark" : "light";
+        this.setPreference();
+      });
+  }
+
+  getValue(): string {
+    return this.themeValue;
+  }
+
+  render(): ReturnType<LitElement["render"]> {
+    return html`
+      <button
+        title="Switch to ${this.themeValue === "light" ? "dark" : "light"} mode"
+        aria-label="auto"
+        aria-live="polite"
+        data-theme="${this.themeValue}"
+        @click="${this.onClick}"
+      >
+        <svg class="sun-and-moon" aria-hidden="true" viewBox="0 0 24 24">
+          <mask class="moon" id="moon-mask">
+            <rect x="0" y="0" width="100%" height="100%" fill="white" />
+            <circle cx="25" cy="10" r="6" fill="black" />
+          </mask>
+          <circle
+            class="sun"
+            cx="12"
+            cy="12"
+            r="6"
+            mask="url(#moon-mask)"
```

Comment:
Something like this to help reduce the chance of id name clashes?

```suggestion
          <mask class="moon" id="_dark-mode-switch-moon-mask_">
            <rect x="0" y="0" width="100%" height="100%" fill="white" />
            <circle cx="25" cy="10" r="6" fill="black" />
          </mask>
          <circle
            class="sun"
            cx="12"
            cy="12"
            r="6"
            mask="url(#_dark-mode-switch-moon-mask_)"
```

Reply:
This is in the shadow DOM, so I don't think we need to worry about name clashes.

Reply:
Ah, good point!